src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/Serialization/JsonDataReader.cs

```diff
@@ -0,0 +1,528 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.Extensions.ObjectPool;
+using Newtonsoft.Json;
+
+namespace Microsoft.AspNetCore.Razor.ProjectEngineHost.Serialization;
+
+internal delegate void ReadPropertyValue<TData>(JsonDataReader reader, ref TData data);
+internal delegate T ReadValue<T>(JsonDataReader reader);
+internal delegate T ReadProperties<T>(JsonDataReader reader);
+internal delegate void ProcessValue<T>(JsonDataReader reader, T arg);
+internal delegate void ProcessProperties<T>(JsonDataReader reader, T arg);
+
+internal sealed class PropertyMap<TData>
```

Comment:
If this isn't going to be file scoped, can we move to a separate file?

Reply:
Sure!